File: 2015/js/day01.js
import '../../utils/js/array'
import '../../utils/js/math'

const symbols = [1, -1]
const floor = s => symbols.pickWhen(s == '(')

const part1 = ops => ops.sum()

const part2 = ops => ops
  .scan(Math.plus, 0)
  .indexOf(-1)

export { floor, part1, part2 }

File: 2015/hs/src/day01.hs
step :: Num t => Char -> t
step '(' = 1
step ')' = -1
step _ = 0

floor :: (Functor f, Num b) => f Char -> f b
floor s = step <$> s

part1 :: [Char] -> Int
part1 s = sum $ floor s

part2 :: [Char] -> Int
part2 s = length $ takeWhile (>=0) $ scanl (+) 0 $ floor s

-- part1 <$> getLine